Add CCZ4 function for the spatial ricci tensor

macedo22 opened the pull request:
## Proposed changes

This PR adds a CCZ4 function for computing the spatial ricci tensor.

The motivation for this PR is to implement eq. (20), whose value will be used to implement eq. (12e) and (27) in [this CCZ4 paper](https://arxiv.org/pdf/1707.09910.pdf), whose value will later be necessary for computing the evolved variables.

Since eq (20) is the only computation that depends on the spatial derivative of the spatial christoffel symbols (identity defined by eq. (18)) and since it only needs select components from the riemann tensor (eq. 19), the computation of eq. (18) and eq. (19) are not defined generically in separate functions. Instead, this implementation of eq. (20) only computes the components it needs from the eq. (18) LHS and the eq. (19) LHS.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

macedo22 commented on src/Evolution/Systems/Ccz4/Ricci.hpp:
```diff
@@ -0,0 +1,71 @@
+ * `Ccz4::Tags::FieldP`, and its spatial derivative, respectively.
+ */
+template <size_t Dim, typename Frame, typename DataType>
+void ricci(
```
Note that for some reason, while the documentation does properly generate for this function, its page does not properly get linked to the website's search result for "ricci".

macedo22 replied:
The `details` math in this documentation is identical to the `details` math in the documentation for the `Ccz4::Ricci` tag, but the link to the `Ccz4::Ricci` tag is properly linked to the search result. Perhaps something is wrong with the `\ingroup`? I've been using `\ingroup GeneralRelativityGroup` for all `Ccz4` functions, and they are properly linked on the website outside of this `ricci` one.

nilsdeppe commented on src/Evolution/Systems/Ccz4/Ricci.hpp:
```diff
@@ -0,0 +1,71 @@
+namespace Ccz4 {
+/// @{
+/*!
+ * \ingroup GeneralRelativityGroup
```
this shouldn't be in the GR group. It's CCZ4-specific while the GR group is generic. This code should be findable by going `Code Reference->Topics->Evolution Systems->Ccz4`. I think that namespace would normally be documented in `Evolution/Systems/Ccz4/System.hpp`. If you want, you can add that file with just documenting the namespace. This would also mean we wouldn't need to use the search functionality to find the documentation 🙃

nilsdeppe commented on src/Evolution/Systems/Ccz4/Ricci.hpp:
```diff
@@ -0,0 +1,71 @@
+ * \brief Computes the spatial ricci tensor
+ *
+ * \details Computes the spatial ricci tensor as:
+ * \f{align}
```
It's best to have a blank line before `\f{align}` and after `\f}` so that clang-format doesn't combine lines and break math rendering.

nilsdeppe commented on src/Evolution/Systems/Ccz4/Ricci.hpp:
```diff
@@ -0,0 +1,71 @@
+ * \details Computes the spatial ricci tensor as:
+ * \f{align}
+ *     R_{ij} &=
+ *       \partial_k \Gamma^m{}_{ij} - \partial_j \Gamma^m{}_{ik} +
```
The indices are broken here. You have no free `k` and `m` on the LHS

nilsdeppe commented on src/Evolution/Systems/Ccz4/Ricci.hpp:
```diff
@@ -0,0 +1,71 @@
+ *         \tilde{\gamma}_{il} \partial_{(k} P_{j)} -
+ *         \tilde{\gamma}_{ij} \partial_{(k} P_{l)})
+ * \f}
+ * \f$\Gamma^k{}_{ij}\f$, \f$\partial_m \tilde{\Gamma}^k{}_{ij}\f$,
```
"where \f$\Gamma^...\f$ is the spatial Christoffel symbol, \f$\partial_m...\f$ is the spetial derivative of the ..." would make reading this a lot easier. Right now I have to match terms between two long lists.

nilsdeppe commented on src/Evolution/Systems/Ccz4/Ricci.hpp:
```diff
@@ -0,0 +1,71 @@
+ * `Ccz4::Tags::FieldP`, and its spatial derivative, respectively.
+ */
+template <size_t Dim, typename Frame, typename DataType>
+void ricci(
```
Could you name this `spatial_ricci_tensor` since `ricci` could mean the spatial or spacetime and tensor or scalar.

nilsdeppe commented on src/Evolution/Systems/Ccz4/Ricci.cpp:
```diff
@@ -0,0 +1,141 @@
+  for (auto& component : (*result)) {
+    component = 0.0;
+  }
+
```
Did you not want to use the tensor expressions for this computation? Up to you :)

macedo22 replied:
The implementation now uses `TensorExpression`s since https://github.com/sxs-collaboration/spectre/pull/3645 brought down the compilation time

nilsdeppe commented on src/Evolution/Systems/Ccz4/Ricci.cpp:
```diff
@@ -0,0 +1,141 @@
+  for (size_t i = 0; i < Dim; i++) {
+    for (size_t j = i; j < Dim; j++) {
+      for (size_t m = 0; m < Dim; m++) {
+        result->get(i, j) +=
```
It would help if you wrote out the equation you're implementing in the documentation. It's much harder to check that the code is correct without knowing what you coded :). Adding comments like `// Add first term of d_m Gamma^m_{ij} and -d_j Gamma^m_{im}` also helps guide the reader with what you've actually implemented

nilsdeppe commented on src/Evolution/Systems/Ccz4/Ricci.cpp:
```diff
@@ -0,0 +1,141 @@
+            d_conformal_christoffel_second_kind.get(m, m, i, j) -
+            d_conformal_christoffel_second_kind.get(j, m, i, m);
+        for (size_t l = 0; l < Dim; l++) {
+          result->get(i, j) +=
```
until the docs renders and I can compare to your math there (and verify your math in the docs) I can't really review this. It's a lot of terms. Could you please:
1. document the `Ccz4` namespace `Evolution/Systems/Ccz4/System.hpp` in a separate commit
2. document the actual equation you've coded in addition to what you have
3. add code comments about what the terms are matching to in the documentation

macedo22 replied:
In the documentation, I forgot to substitute `m` in for `k` in:
```
 * \f{align}
 *     R_{ij} &=
 *       \partial_k \Gamma^m{}_{ij} - \partial_j \Gamma^m{}_{ik} +
 *       \Gamma^l{}_{ij} \Gamma^m{}_{lk} - \Gamma^l{}_{ik} \Gamma^m{}_{lj}
 * \f}
```

Outside of this definition (once fixed) and the accompanying definition I provided below it for `\partial_k \Gamma^m{}_{ij}` (eq. 18, used for the first two terms in the eq. above), when you say "document the actual equation you've coded in addition to what you have," do you mean that you want me to also write out what `R_{ij}` would be with the all of the terms from eq. 18 substituted (twice) for the first two terms? I originally didn't do this because I thought showing the full substitution might be overkill, so I want to make sure I'm interpreting your request correctly.

nilsdeppe replied:
Yes, I think it would be good to write it out. Someone 5 years from now will look at this code and the more help you can give them the better :)

macedo22 commented:
@nilsdeppe I believe I've addressed all current review requests!

nilsdeppe commented on src/Evolution/Systems/Ccz4/Tags.hpp:
```diff
@@ -193,6 +193,78 @@ struct ChristoffelSecondKind : db::SimpleTag {
 };
 
+/*!
+ * \brief The spatial ricci tensor
```
I don't think copy-pasting the documentation is sustainable. That's a _huge_ amount of duplicate code. Can you either use `\copydoc` or just say `See Ccz4::spatial_ricci_tensor() for details`? This file is already a lot of copy-pasted dox, and nobody will update both places when they update the dox.

nilsdeppe commented on src/Evolution/Systems/Ccz4/Ricci.hpp:
```diff
@@ -0,0 +1,103 @@
+ *
+ * \f{align}
+ *     R_{ij} &=
+ *       \partial_m \Gamma^m{}_{ij} - \partial_j \Gamma^m{}_{im} +
```
The Christoffel symbols are normally written as `\Gamma^m_{ij}`. They are the one of the few things where the indices aren't separated out and it's okay/correct. This needs to also get updated in the cpp file comments

nilsdeppe commented on src/Evolution/Systems/Ccz4/Ricci.cpp:
```diff
@@ -0,0 +1,150 @@
+                          field_d.get(j, i, m) * field_p.get(l))) -
+              // Add terms of \partial_m \Gamma^m{}_{ij} and
+              // -\partial_j \Gamma^m{}_{im} that have a coefficient of 1/2
+              0.5 *
```
there are no terms with 1/2 in your documentation (not explicitly). How about `Add \partial_{(i} P_{j)} type terms`?

nilsdeppe commented on src/Evolution/Systems/Ccz4/Ricci.cpp:
```diff
@@ -0,0 +1,150 @@
+              // -\partial_j \Gamma^m{}_{im} that have a coefficient of 1/2
+              0.5 *
+                  (inverse_conformal_spatial_metric.get(m, l) *
+                   (conformal_spatial_metric.get(j, l) * d_field_p.get(m, i) +
```
why not write this as `conformal_spatial_metric.get(j, l) * (d_{m,i} + d_{i, m})`? You're doing a _lot_ of extra math in these terms that you can eliminate analytically.

nilsdeppe commented on tests/Unit/Evolution/Systems/Ccz4/Ricci.py:
```diff
@@ -0,0 +1,48 @@
+        inverse_conformal_spatial_metric, field_d, field_d_up, field_p,
+        d_field_p)
+
+    return np.einsum(
```
I don't understand what's going on with einsum here. I was expecting something like `einsum("kmij->mmij", d_christoffel_second_kind) + einsum("jmim", d_crhistoffel_second_kind) +...`

macedo22 replied:
This is intended to be the trace of the Riemann tensor (eq. 19 [here](https://arxiv.org/pdf/1707.09910.pdf)) where it's computing `R^m{}_{imj}` since `R_{ij} = R^m{}_{imj}`. The four terms in eq. 19 are ordered `mikj` to align with the ordering for the Riemann tensor, and then the outermost `einsum` computes the trace by using `mimj`.

If this seems correct but was difficult to follow, let me know if you would like me to add comments for this.

If this is incorrect, let me know that, as well!

macedo22 replied:
I guess there is also no reason I can't remove the outer `einsum` and just do something like:
```
    return (np.einsum("mmij", d_christoffel_second_kind) -
         np.einsum("jmim", d_christoffel_second_kind) +
         np.einsum("lij,mlm", christoffel_second_kind,
                   christoffel_second_kind) -
         np.einsum("lim,mlj", christoffel_second_kind,
                   christoffel_second_kind))
```

Let me know if you would like me to change it to something like this ^

nilsdeppe replied:
I find the 4 sums much easier to read than what you currently have since it matches what it's in the dox :)

nilsdeppe commented on tests/Unit/Evolution/Systems/Ccz4/Test_Ricci.cpp:
```diff
@@ -0,0 +1,192 @@
+      determinant_and_inverse(spatial_metric).second;
+
+  // Compute arguments for `spatial_ricci_tensor` function to test
+  const auto conformal_factor = pow(get(det_spatial_metric), -1.0 / 6);
```
`6->6.` and 1.0->1. to be consistent within the file

nilsdeppe commented on tests/Unit/Evolution/Systems/Ccz4/Test_Ricci.cpp:
```diff
@@ -0,0 +1,192 @@
+      conformal_spatial_metric, inverse_conformal_spatial_metric, field_d,
+      field_d_up, field_p, d_field_p);
+
+  CHECK_ITERABLE_APPROX(expected_ricci_tensor, actual_ricci_tensor);
```
can you also compare to what you get when you call `gr::ricci_tensor`? That provides another independent check. Currently if you have the same error in your python and C++ code the test will still pass

macedo22 commented:
@nilsdeppe Ready for rereview! My original 3 commits have different hashes now because I had to rebase to be able to build for the CI. The fixup commit should have the only new changes that address your most recent review's requests.

nilsdeppe commented on tests/Unit/Evolution/Systems/Ccz4/Test_Ricci.cpp:
```diff
@@ -155,23 +162,49 @@ void test_compute_spatial_ricci_tensor(
-  CHECK_ITERABLE_APPROX(expected_ricci_tensor, actual_ricci_tensor);
+  CHECK_ITERABLE_APPROX(expected_python_ricci_tensor, actual_ricci_tensor);
+
+  Approx approx = Approx::custom().epsilon(1e-9).scale(1.0);
```
not a change request, I'm just surprised the error is at 1e-9 rather than 1e-13 level. Do you understand why? This is sort of suspicious

macedo22 replied:
I don't know why. However, I wouldn't really say I have a baseline model in my head of how far off it should be given that numerical derivatives were used. I did check that intermediate tensors (approximately) agreed with expected, and I believe those required either 1e-10 or 1e-12 error. Would you like me to follow up on this computation with more tests? Or I can add back in the `CHECK`s for the intermediate values in this test. Let me know!

nilsdeppe replied:
It could just be the numerical derivatives. If you increase the number of grid points does the error decrease? E.g. if you use `num_pts_1d = 10` (we haven't really used the word `grid_size` for this since the "size" is really a physical thing). If the test takes to long to run with 10 or 11 points that's fine, it would be good to have a comment saying why you need such a comparatively large tolerance

macedo22 replied:
Interestingly enough, when I bump it up to `grid_size = 12`, I end up having to weaken the approx requirement to `epsilon = 1e-8`. I'm assuming it's best to go back over this and see if there is something incorrect about the test, but let me know if you feel otherwise.

macedo22 replied:
P.S. bumping up to `grid_size = 10` from `8` had no effect on the required `epsilon` being `1e-9`

nilsdeppe replied:
I would print the l2norm (Blaze has support for this) of the difference `expected-computed` for grid size 2, 4, 6, 8, 10. Hopefully you see some sort of convergence :)

macedo22 replied:
Oh, is it not unusual for it to not monotonically decrease? The L2 norm seems to be converging on a minimum for `num_points_1d = 6`. I would have naively expected it to monotonically decrease as grid points increased just because I'm thinking more grid points => more accurate.

When `num_points_1d = 6` is used, the test passes at best with `epsilon = 1e-10`.

nilsdeppe replied:
Well it should monotonically decrease, but at some point things like rounding errors etc. will dominate. However, that point is usually around 1e-12 to 1e-14. This depends on size of the domain as well. There are a number of things that can break convergence, so not obvious what the issue is. It might also be that both things are converging, just to different values. That's generally bad, but if it's converging to different values that can also be dominated by rounding errors or just how the equations are evaluated. But again, normally that's 1e-12 or 1e-14 level

macedo22 replied:
Reporting back:

Dr. Lovelace noticed that my coordinates for this test were inside the horizon and suggested I make the coordinates farther away. Here are the parameters I originally used (on this PR):
```
  const double mass = 2.;
  const std::array<double, 3> spin{{0.3, 0.5, 0.2}};
  const std::array<double, 3> center{{0.2, 0.3, 0.4}};
  const size_t grid_size = 8;
  const std::array<double, 3> lower_bound{{0.82, 1.24, 1.32}};
  const std::array<double, 3> upper_bound{{0.8, 1.22, 1.3}};
```
And when I increase the bounds values by a factor 10 to:
```
  const std::array<double, 3> lower_bound{{8.2, 12.4, 13.2}};
  const std::array<double, 3> upper_bound{{8.0, 12.2, 13.0}};
```
the test passes with `1e-11` at best instead of `1e-9`. Note that, as before, increasing the number of grid points just makes the difference worse.

Do you have thoughts about this?

nilsdeppe replied:
Sounds like one of (or both) of the ways you're calculating things has bad cancellation errors. This will effectively mean that the CCZ4 system would ever be accurate beyond 1e-9, which isn't great. Keep in mind that with CCZ4 you're simulating inside the horizon so moving further out isn't a luxury you have in this case. I still find this worrying. I think the KS solution has an analytic expression for the spatial christoffel symbols that you could use instead of numerical derivatives to gain some insight into which expression is bad.

macedo22 replied:
Just to add to the story in case it's relevant, while I was "backing up" through the computation of the ricci tensor, I also checked to see how closely the intermediate values matched what was expected to try to see how early on in the computation of the ricci tensor that calculations differed significantly.

(1) Taking the LHS tensor for eq (15) in the paper and using `partial_derivatives` to compute the numerical derivative vs. computing it analytically using eq (16) differ by `1e-11` as well. This happens even if the symmetrization in eq (16) is not implemented.

(2) These methods of calculating the derivative of the conformal metric differ by `1e-11`:
- Kerr spatial metric extracted -> multiply by conformal factor to get conformal spatial metric -> `partial_derivatives` to get numerical derivative of conformal metric
- Kerr spatial metric and Kerr d_spatial_metric extracted -> `partial_derivatives` to get spatial derivative of the square of the conformal factor -> analytical equation for the spatial derivative of the conformal metric: `\partial_k \tilde{\gamma}_{ij} = \phi^2 \partial_k \gamma_{ij} + \partial_k \phi^2 \gamma_{ij}`

macedo22 replied:
Substituting in alternatively-calculated intermediates doesn't fix the downstream numerical error with the ricci tensor. This is just meant to say that it seems like I'm seeing the less-than-optimal error with other calculations' comparisons involving numerical derivatives.

nilsdeppe reviewed (commented):
LGTM, please squash

macedo22 commented:
@nilsdeppe Squashed! I think either this or https://github.com/sxs-collaboration/spectre/pull/3579 will need to be merged first and the other will need to be rebased after.

nilsdeppe commented:
Does this need to be rebased because of #3579? There don't appear to be direct conflicts, but I'm not sure what you were expecting to conflict

macedo22 commented:
I haven't yet rebased but I expect conflicts due to the first commit a093290 in things like `Tags.hpp` and `Test_Tags.cpp`. The reason I hadn't yet rebased is because although you initially mentioned that the numerical error we're seeing in the test was not a change request, the conversation that followed seemed to imply that we do need to get to the bottom of the numerical error before this PR is merged. I'll go ahead and rebase now but let me know if I am correct/incorrect in thinking that finding the source of the numerical error needs to precede merging.

nilsdeppe commented:
Okay, thanks! I had forgotten about the errors when I posted my question about rebasing.

@nilsleiffischer I think the errors here are possibly related to what you're seeing in the elliptic solver. This is computing the Ricci tensor in Kerr. The resulting errors were ~1e-9 when computed two different ways, which I found to be surprisingly large. Not as a large as what you have, but might be related. Any thoughts on this here?

macedo22 commented:
@nilsdeppe Turns out there are no conflicts to resolve or unwanted results from auto-merging, so I'm choosing not to rebase :+1:

macedo22 commented:
Requires https://github.com/sxs-collaboration/spectre/pull/3610 (Kerr-Schild analytical computation of the derivative of the determinant of the spatial metric) to be merged first.

macedo22 commented:
@nilsdeppe I've rebased and left the original commit that added the Ricci function and tag (with small change to capitalize `Ricci` in documentation) so it's easy to see what was previously reviewed. Everything done in the fixup is tweaks I made to improve the error from `1e-9` to `1e-12`, which involves swapping in more analytical derivatives for values in the calculation of the expected Ricci tensor from `gr`.

nilsdeppe commented on tests/Unit/Evolution/Systems/Ccz4/Test_Ricci.cpp:
```diff
@@ -89,19 +91,24 @@ void test_compute_spatial_ricci_tensor(
-  const auto& d_conformal_spatial_metric =
-      get<Tags::deriv<conformal_spatial_metric_tag, tmpl::size_t<SpatialDim>,
-                      Frame::Inertial>>(d_conformal_spatial_metric_var);
+  gr::Solutions::KerrSchild::IntermediateVars<DataVector, Frame::Inertial>
```
Can you do `const auto d_det.. = get<gr::Tags::DerivDetSpatialMetric<...>>(solution.variables(x, t, tmpl::list<gr::Tags::DerivDetSpatialMetric<...>>{}));` instead of relying on the intermediate vars?

macedo22 replied:
I could be wrong, but to be able to do this, wouldn't `DerivDetSpatialMetric` need to be included in [this list](https://github.com/sxs-collaboration/spectre/blob/fc92aea9d9c6403fdf0aab2d4859f76c95d8258b/src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Solutions.hpp#L29) that is common to all GR solutions? I didn't add it there when I added it to KerrSchild [here](https://github.com/sxs-collaboration/spectre/pull/3610).

nilsdeppe replied:
ah... I'm not sure it needs to be in the last, I didn't write the KS solution. However, I think the tag should be retrievable from `variables`. The `IntermediateVars` is really an implementation detail, not the way you should get variables from the solution...

macedo22 replied:
If I confirm that it does, do you feel it's appropriate to add this tag as something that is common to all of the Analytic GR solutions?

nilsdeppe replied:
It might be useful, but I don't think it should be necessary to add everywhere. That would be a rather restrictive design and we should figure out why it seems to be needed everywhere.

macedo22 replied:
Sorry, I jumped the gun with that question. I can confirm whether or not this is the case first, and then let you know

macedo22 replied:
Yes, this does appear to be the case. [This static assert](https://github.com/sxs-collaboration/spectre/blob/fc92aea9d9c6403fdf0aab2d4859f76c95d8258b/src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp#L256) seems to be failing because `DerivDetSpatialMetric` is not a tag in `tags<DataType, Frame>`, which is defined [here](https://github.com/sxs-collaboration/spectre/blob/fc92aea9d9c6403fdf0aab2d4859f76c95d8258b/src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Solutions.hpp#L29), which seems to be a list common to all GR solutions.

Let me know how you would like to proceed! Not sure if this is a design thing you want me or someone else to change, or if you want me to just compute it with the intermediate for now.

nilsdeppe replied:
does it work if you change the assert to use `tmpl::push_back<tags<DataType, Frame>, DerivDetSpatialMetric>` (the extra tag)?

macedo22 replied:
Yes, that does work. It makes me wonder if there was a reason that when this `variables` function was written, there was a reason to not allow intermediates to be retrieved in this way

nilsdeppe replied:
You don't need access to most of the intermediate quantities. In general you only need access to the 3+1 quantities. You don't need access to H, l_a, and all the other intermediate things.

macedo22 replied:
Ah yes. Did you want me to make the change to the static assert that you suggested? In light of the fact that intermediates aren't things you typically want to retrieve, I had thought you were just asking me to do that change to the static assert to confirm that that was the problem. Let me know if this is a change you'd like me to make as part of this PR.

nilsdeppe replied:
Yes, please make the change to the assert. Sorry for not being clear on that!

macedo22 replied:
Also my bad! Will push the fix soon

nilsdeppe reviewed (commented):
LGTM, just one small suggestion. You can squash, including any additional changes :)

macedo22 committed:
Make DerivDetSpatialMetric tag retrievable from KerrSchild

macedo22 committed:
Add CCZ4 function and tag for the spatial Ricci tensor

macedo22 commented:
@nilsdeppe This is ready for review. Two things to note:

- I added a new first commit that changes the static assert in `KerrSchild` to allow us to retrieve `DerivDetSpatialMetric`
- I changed the `spatial_ricci_tensor` function to use `TensorExpression`s instead of loops because after https://github.com/sxs-collaboration/spectre/pull/3645, the compilation time for the `DataVector` version of the TE has been brought down to ~10s, and I figure it's good to have a real equation we need that this nasty written with TEs to see how it does

macedo22 commented:
Oh shoot, actually, I just realized that now that the unit test here tests against `KerrSchild` instead of a python function, the python test functions file (`Ricci.py`) is not needed. Let me know if you would like me to add back the python test, and I can push a squashed fix to `Test_Ricci.cpp`. If not, I will push a squashed fix to remove `Ricci.py` and the `pypp` stuff from `Test_Ricci.cpp` and _then_ it will be ready :upside_down_face:

nilsdeppe commented:
I think generally we prefer to have both python and analytic tests. So if that's possible without too much work, that would be great :)

macedo22 commented:
@nilsdeppe Ok, it is now ready for review. I've added back the python test. It was originally not written as a separate test function because I believe that random input tensors for the python function would lead to a test failure because of symmetries with the ricci tensor, i.e. I kept the test as it was, using KS input values, which should manifest any symmetry properties necessary for the function to pass.

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp:
```diff
@@ -252,8 +252,11 @@ class KerrSchild : public AnalyticSolution<3_st>,
     static_assert(
-        tmpl2::flat_all_v<
-            tmpl::list_contains_v<tags<DataType, Frame>, Tags>...>,
+        tmpl2::flat_all_v<tmpl::list_contains_v<
```
can you make sure the KS test doesn't retrieve via the intermediate vars?

macedo22 replied:
I'm not sure I follow :/ [Here](https://github.com/sxs-collaboration/spectre/blob/7b8e2062f46dd6e71331b5930511f58aef2cd8c1/tests/Unit/Evolution/Systems/Ccz4/Test_Ricci.cpp#L79) it's retrieved without using intermediate vars, but my understanding is that `KerrSchild` will still use it internally. Let me know if you think I'm misunderstanding your request!

nilsdeppe replied:
Sorry, the call https://github.com/sxs-collaboration/spectre/blob/7b8e2062f46dd6e71331b5930511f58aef2cd8c1/tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Test_KerrSchild.cpp#L208 should not go through the intermediate vars is what I'm saying. Does that clarify things?

macedo22 replied:
Oh I see, I definitely misunderstood. I will make that fix now. Thanks for clarifying!

macedo22 replied:
@nilsdeppe I've squashed in this fix

nilsdeppe reviewed (commented):
Looks good, just one question. Please squash if there are any changes necessary

macedo22 commented:
Merging this one gets a 🎉
